extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTools.java

```diff
@@ -220,21 +221,122 @@ public static void writeTable(@NotNull final Table sourceTable,
+        final File firstCreated = prepareDestinationFileLocation(shadowDestFile);
+        ArrayList<File> cleanupFiles = new ArrayList<>();
+        ArrayList<File> rollbackFiles = new ArrayList<>();
         try {
-            writeParquetTableImpl(
-                    sourceTable, definition, writeInstructions, destFile, definition.getGroupingColumnNamesArray());
+            Map<String, ParquetTableWriter.GroupingColumnWritingInfo> groupingColumnWritingInfoMap = null;
+            final String[] groupingColumnNames = definition.getGroupingColumnNamesArray();
+            if (groupingColumnNames.length > 0) {
+                // Create a map from grouping column name to writing info (like paths, etc.) for passing down to lower
+                // layers and delete any existing backup files for these grouping columns
+                final String[] parquetColumnNames = Arrays.stream(groupingColumnNames)
+                        .map(writeInstructions::getParquetColumnNameFromColumnNameOrDefault)
+                        .toArray(String[]::new);
+                groupingColumnWritingInfoMap =
+                        groupingColumnInfoBuilderHelper(groupingColumnNames, parquetColumnNames, destFile);
+            }
+
+            // Store all the files for cleanup in case of exception
+            cleanupFiles.add(shadowDestFile);
+            if (groupingColumnWritingInfoMap != null) {
+                for (final ParquetTableWriter.GroupingColumnWritingInfo gcwi : groupingColumnWritingInfoMap.values()) {
+                    cleanupFiles.add(gcwi.destFile);
+                }
+            }
+
+            ParquetTableWriter.write(sourceTable, definition, writeInstructions, shadowDestFile.getPath(),
+                    Collections.emptyMap(), groupingColumnWritingInfoMap);
+
+            // Write to shadow files was successful
+            installShadowFile(destFile, shadowDestFile);
+            rollbackFiles.add(destFile);
+            if (groupingColumnWritingInfoMap != null) {
+                for (final ParquetTableWriter.GroupingColumnWritingInfo gfi : groupingColumnWritingInfoMap.values()) {
+                    final File groupingDestFile = gfi.metadataFilePath;
+                    final File shadowGroupingFile = gfi.destFile;
+                    installShadowFile(groupingDestFile, shadowGroupingFile);
+                    rollbackFiles.add(groupingDestFile);
+                }
+            }
         } catch (Exception e) {
+            for (final File file : rollbackFiles) {
+                rollbackFile(file);
+            }
+            for (final File file : cleanupFiles) {
+                file.delete();
+            }
             if (firstCreated != null) {
                 FileUtils.deleteRecursivelyOnNFS(firstCreated);
-            } else {
-                // noinspection ResultOfMethodCallIgnored
-                destFile.delete();
             }
-            throw e;
```

Comment:
I don't understand why we didn't have to declare this in `throws`. Your existing version is better.